sematic/graph.py

```diff
@@ -485,55 +515,54 @@ def _get_run_output(self, run_id: RunID) -> Tuple[Any, Optional[Artifact]]:
-        reset_run_ids: List[RunID],
-    ):
+        future_graph: FutureGraph,
+    ) -> None:
+        """Use this run graph and provided run to set the parent future in future_graph.
+
+        To be used in the process of constructing a FutureGraph from this run graph
+        (potentially as a cloned future graph). This method will update futures in
+        future_graph in-place.
+
+        The parent run of the provided run will be used to identify the appropriate
+        parent future for the provided future. The provided future and its parent
+        future will be modified as necessary.
+
+        Parameters
+        ----------
+        future:
+           The future whose parent should be updated.
+        run:
+            The run to be used as a reference to update the future's parent.
+        future_graph:
+            The future graph being constructed.
+        """
         if run.parent_id is None:
             return
 
-        parent_future = cloned_graph.futures_by_original_id[run.parent_id]
+        parent_future = future_graph.futures_by_run_id[run.parent_id]
 
         future.parent_future = parent_future
 
         # Is future parent_future's nested future?
         if self._runs_by_id[run.parent_id].nested_future_id == run.id:
             parent_future.nested_future = future
 
-            if run.id in reset_run_ids:
-                parent_future.state = FutureState.RAN
-
     def _clone_future(
         self, run_id: RunID, cloned_graph: ClonedFutureGraph, reset_run_ids: List[RunID]
     ) -> AbstractFuture:
         run = self._runs_by_id[run_id]
 
-        kwargs, run_input_artifacts = self._get_cloned_future_inputs(
-            run_id, cloned_graph
-        )
-
-        func = run.get_func()
-
-        # _make_list and _make_tuple need special treatment as they are not
-        # decorated functions, but factories that dynamically generate futures.
-        if run.function_path == "sematic.function._make_list":
-            # Dict values insertion order guaranteed as of Python 3.7
-            input_list = list(kwargs.values())
-            future = func(make_list_type(input_list), input_list)  # type: ignore
-        elif run.function_path == "sematic.function._make_tuple":
-            # Dict values insertion order guaranteed as of Python 3.7
-            input_tuple = tuple(kwargs.values())
-            future = func(make_tuple_type(input_tuple), input_tuple)  # type: ignore
-        else:
-            future = func(**kwargs)
+        kwargs, run_input_artifacts = self._get_future_inputs(run_id, cloned_graph)
 
-        future.name = run.name
+        future = initialize_future_from_run(run, kwargs, use_same_id=False)
 
-        future.props.name = run.name
-        future.props.tags = json.loads(str(run.tags))
+        # For cloning a graph, we want the default state of the future to be
+        # created unless explicitly updated to a different value.
+        future.props.state = FutureState.CREATED
```

Comment:
No, `initialize_future_from_run` will use the `future_state` from the run as the `state` for the `future`. In the case of runner reentrancy, this is what is desired. In the case of a graph clone, the future should be in the `CREATED` state unless specified otherwise.